File: Sorting/Sorting/RankingSystem.cs
namespace Sorting
{
	using System.Collections.Generic;
	using System.Linq;

	public class RankingSystem
	{
		private readonly List<Position> _ranking = new List<Position>();
		private readonly int _sortingsCount;

		public RankingSystem(int sortingsCount)
		{
			_sortingsCount = sortingsCount;
		}

		public void Clear()
		{
			_ranking.Clear();
		}

		public void Add(string sortName, long duration)
		{
			_ranking.Add(new Position(sortName, duration));
		}

		public int? Get(string sortName)
		{
			if (_ranking.Count != _sortingsCount)
			{
				return null;
			}

			var rank = _ranking
				.OrderBy(position => position.Duration)
				.Select((position, index) => new { P = position, Rank = index })
				.FirstOrDefault(o => o.P.Name == sortName);

			if (rank == null)
			{
				throw new KeyNotFoundException($"Sort name '{sortName}' not found");
			}

			return rank.Rank + 1;
		}

		private class Position
		{
			public Position(string name, long duration)
			{
				Name = name;
				Duration = duration;
			}

			public string Name { get; }

			public long Duration { get; }
		}
	}
}

File: Sorting/Sorting.WebPure/scripts/Tools/RankingSystem.ts
namespace Tools {
	export class RankingSystem {
		private ranking: Candidate[] = [];

		public Clear(): void {
			this.ranking = [];
		}

		public Add(sortName: string, duration: number): void {
			this.ranking.push(new Candidate(sortName, duration));
		}

		public Get(sortName: string): number {
			const rank = this.ranking
				.sort((a: Candidate, b: Candidate) => (a.duration - b.duration))
				.map((candidate, index) => new Rank(candidate, index ))
				.filter(r => r.candidate.name === sortName)
				[0];

			if (!rank) {
				throw `Sort name '${sortName}' not found`;
			}

			return rank.rank + 1;
		}
	}

	class Rank {
		candidate:Candidate;
		rank: number;
		constructor(candidate: Candidate, rank: number) {
			this.candidate = candidate;
			this.rank = rank;
		}
	}

	class Candidate {
		name: string;
		duration: number;
		constructor(name: string, duration: number) {
			this.name = name;
			this.duration = duration;
		}
	}
}